feat(core/remio): add Remote I/O infrastructure

joaopeixoto13 opened the pull request:
This PR introduces the foundational support for VirtIO by implementing a mechanism to forward VirtIO requests bidirectionally between the guest VM (or frontend VM) and its associated backend VM.

**Validation**
- Platforms
  - Armv8-A AArch64
    - Xilinx Zynq UltraScale+ MPSoC ZCU102/4
    - NVIDIA Jetson TX2
    - QEMU virt
  - RISC-V RV64
    - QEMU virt
- System configuration
  - Multiple VirtIO frontend drivers on multiple guest VMs, each running on multiple vCPUs
  - Multiple VirtIO backend devices on multiple backend VMs, each running on multiple vCPUs
- Backend devices
  - Rust-VMM backends ([VirtIO-Block](https://github.com/rust-vmm/vm-virtio/tree/main/virtio-blk), [VirtIO-Net](https://github.com/rust-vmm/vmm-reference/pull/49), [VirtIO-Console](https://github.com/rust-vmm/vm-virtio/tree/main/virtio-console))
  - QEMU (VirtIO-Block, VirtIO-Net, VirtIO-Console)
  - Vhost-user backends ([vhost-user-fs](https://gitlab.com/virtio-fs/virtiofsd), [vhost-user-vsock](https://github.com/rust-vmm/vhost-device/tree/main/vhost-device-vsock))
  - Vhost backends (vhost-vsock)

DavidMCerdeira commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,513 @@
+    }
+}
+
+void remote_io_assign_cpus(struct vm* vm)
```
change to `remote_io_assign_vm_cpus`

DavidMCerdeira commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,513 @@
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns true if the operation was successful, false otherwise
+ */
+static bool remote_io_w_r_operation(unsigned long id, unsigned long addr, unsigned long value,
```
change to `remote_io_rw_operation`

DavidMCerdeira commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,513 @@
+        return false;
+    }
+
+    list_foreach (remote_io_device_list, struct remote_io_device, io_device) {
+        if (io_device->id == dev.id) {
```
separate finding the device from the rest
```
struct *io_device = remote_io_find_io_dev_by_id(...)`
if(!io_device) {
  ERROR(...)
}
...
 ```

DavidMCerdeira reviewed (changes requested):
Generic comment: the namespace for this module can be shortened. I suggest rmio, remio, rm_io, rem_io

some of the comments may be a bit subjective. @josecm feel free to wage in

DavidMCerdeira commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,513 @@
+        struct vm_config* vm_config = &config.vmlist[vm_id];
+        for (i = 0; i < vm_config->platform.remote_io_dev_num; i++) {
+            struct remote_io_dev* dev = &vm_config->platform.remote_io_devs[i];
+            if (devices[dev->id][0] != REMOTE_IO_DEVICE_UNINITIALIZED && dev->is_backend) {
```
on second though, not using bool
But using enum it would work

joaopeixoto13 commented:
This new version includes a refactor of the code following @DavidMCerdeira's suggestions. 
The key changes are as follows:
- Introduced helper functions as part of a new API, such as `remote_io_create_request`, `remote_io_get_request`, `remote_io_push_request_event`, `remote_io_pop_request_event`, etc.
- Decoupled the large `remote_io_hypercall` function into smaller, modular functions, each corresponding to a specific hypercall event.
- Added the `REMOTE_IO_DEV_TYPE` enum for better type safety and code clarity.
- Introduced a new type, `remote_io_id_t`, to manage the Remote I/O ID more effectively.

This refactoring improves the structure and maintainability of the code.

joaopeixoto13 commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,513 @@
+            if (node->addr != addr || node->state != REMOTE_IO_STATE_PROCESSING) {
+                break;
+            }
+
```
The lock is only required to prevent the backend VM from extracting the I/O requests from the slot while the frontend VM is still inserting them. However, this race condition only occurs when the backend VM operates in polling mode. When driven by interrupts, we ensure that the request field is fully updated in the I/O request array slot before raising the interrupt.

Since the I/O request slot (or node) is managed via a finite state machine (FSM) and updated sequentially, no lock is necessary. The design ensures there is no concurrent access to the same variable, even when the frontend and backend are running on multiple vCPUs.

joaopeixoto13 commented:
> Generic comment: the namespace for this module can be shortened. I suggest rmio, remio, rm_io, rem_io
> 
> some of the comments may be a bit subjective. @josecm feel free to wage in

@josecm What are your thoughts on this? From @DavidMCerdeira's list of suggestions, `remio` seems like the most appropriate name. However, I'm unsure if `remote_io` is still considered too lengthy. For this reason, I've retained the current name in the latest version.

joaopeixoto13 commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,513 @@
+            struct remote_io_request* node = &remote_io_requests[cpu_id][vcpu_id];
+            spin_unlock(&remote_io_request_lock);
+
+            if (node->addr != addr || node->state != REMOTE_IO_STATE_PROCESSING) {
```
In fact, address verification is not required at this stage. I initially used it for debugging purposes to confirm that the backend VM was completing requests with the correct MMIO address. However, in the final implementation, this validation is unnecessary. We only need the CPU and vCPU IDs to retrieve the appropriate I/O request slot and the value for read operations. In the post-work job, the accessed register can then be updated with the value provided by the backend.

josecm commented:
> > Generic comment: the namespace for this module can be shortened. I suggest rmio, remio, rm_io, rem_io
> > some of the comments may be a bit subjective. @josecm feel free to wage in
> 
> @josecm What are your thoughts on this? From @DavidMCerdeira's list of suggestions, `remio` seems like the most appropriate name. However, I'm unsure if `remote_io` is still considered too lengthy. For this reason, I've retained the current name in the latest version.

My suggestion would be either `remio` or simply `rio`.

joaopeixoto13 commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,688 @@
+    }
+    struct remote_io_request* request = remote_io_get_request(cpu_id, vcpu_id);
+
+    switch (event) {
+        case REMOTE_IO_CPU_MSG_READ:
+            vcpu_writereg(cpu()->vcpu, request->reg, request->value);
+            break;
+        default:
+            break;
+    }
```
Since the I/O request slot is managed by a finite state machine (FSM) and updated sequentially by both the frontend and backend VMs, locking is unnecessary. To break it down:

1. The request is created by the frontend VM during the MMIO emulation handler (`remote_io_mmio_emul_handler`).

From that point on, the request slot is updated sequentially by the frontend and backend VMs. Specifically:

2. During the ask operation (`remote_io_handle_ask`), the backend updates the request state.
3. Once the backend finishes processing the request, it updates both the state and value (`remote_io_handle_rw`).
4. Finally, the frontend VM updates the state during the post-work phase (`remote_io_cpu_post_work`).

This ensures that:

1. The backend and frontend VMs will never access or modify the request simultaneously. When the backend VM operates in polling mode, it checks for pending requests by popping a request event, which is protected by the list lock, rather than directly inspecting the request. In interrupt mode, the interrupt is only injected after the I/O request slot has been fully updated. As in polling mode, the backend checks for new pending requests by examining the event list. Only when an event is present (indicating that the frontend has already inserted the request into the I/O request array via `remote_io_push_request_event` after `remote_io_insert_request`) will the backend fetch the I/O request slot.
2. If the frontend VM runs on multiple vCPUs, each vCPU can only make one request at a time. Since the vCPU goes into an idle state after issuing a request, it is impossible for multiple vCPUs to access the same slot simultaneously.
3. If the backend VM runs on multiple vCPUs, each vCPU handling the hypercall will be responsible for processing only one request slot at a time, given by the CPU and vCPU IDs.

Although I have tested the infrastructure on real hardware platforms (e.g., ZCU102), running both frontend and backend VMs on multiple vCPUs without encountering any issues, please feel free to highlight any potential corner cases that I may have missed at this stage.

DavidMCerdeira replied:
Thanks for the explanation I think I understand
However, I fail to understand in what scenarios is the lock actually needed if the vcpu will always for a request to be completed before issuing another

joaopeixoto13 replied:
Yes, you are correct! A global lock is unnecessary, as in the ask routine, we only fetch the I/O slot from the array when there are pending I/O requests in the event request list. In fact, I realized this while drafting the message above. :)

In earlier versions, I implemented the lock because there was no event request list; therefore, the verification and the actual retrieval process were performed on the same object.

joaopeixoto13 commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,688 @@
+
+    for (size_t i = 0; i < vm->remote_io_dev_num; i++) {
+        dev = &vm->remote_io_devs[i];
+        if (addr >= vm->remote_io_devs[i].va &&
```
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/880ec0feeecee69b8fb6ceb880e31dbaf191fb7e

joaopeixoto13 commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,688 @@
+void remote_io_init(void)
+{
+    size_t frontend_cnt = 0, backend_cnt = 0;
+    int devices[REMOTE_IO_MAX_DEVICES][REMOTE_IO_DEV_BACKEND - REMOTE_IO_DEV_FRONTEND + 1];
```
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/c679c4ad42907ae56732ec6c4b51b8a62d299d68

joaopeixoto13 commented on src/core/inc/remote_io.h:
```diff
@@ -0,0 +1,87 @@
+    size_t size;                        /**< Frontend MMIO size */
+    irqid_t interrupt;                  /**< Frontend/backend interrupt number */
+    remote_io_id_t id;                  /**< Remote I/O ID */
+    enum REMOTE_IO_DEV_TYPE is_backend; /**< True if the device is a backend */
```
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/3b9ff100821a9d8d9737eccc7c48e38bfbf8b52b

joaopeixoto13 commented on src/core/remote_io.c:
```diff
@@ -0,0 +1,688 @@
+    for (size_t i = 0; i < vm->remote_io_dev_num; i++) {
+        dev = &vm->remote_io_devs[i];
+        if (addr >= vm->remote_io_devs[i].va &&
+            addr <= vm->remote_io_devs[i].va + vm->remote_io_devs[i].size) {
```
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/880ec0feeecee69b8fb6ceb880e31dbaf191fb7e

joaopeixoto13 commented:
> > > Generic comment: the namespace for this module can be shortened. I suggest rmio, remio, rm_io, rem_io
> > > some of the comments may be a bit subjective. @josecm feel free to wage in
> > 
> > 
> > @josecm What are your thoughts on this? From @DavidMCerdeira's list of suggestions, `remio` seems like the most appropriate name. However, I'm unsure if `remote_io` is still considered too lengthy. For this reason, I've retained the current name in the latest version.
> 
> My suggestion would be either `remio` or simply `rio`.

Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/01c2cbed6e573c28af1b82229e798d1298da16a6

DavidMCerdeira commented:
> > > > Generic comment: the namespace for this module can be shortened. I suggest rmio, remio, rm_io, rem_io
> > > > some of the comments may be a bit subjective. @josecm feel free to wage in
> > > 
> > > 
> > > @josecm What are your thoughts on this? From @DavidMCerdeira's list of suggestions, `remio` seems like the most appropriate name. However, I'm unsure if `remote_io` is still considered too lengthy. For this reason, I've retained the current name in the latest version.
> > 
> > 
> > My suggestion would be either `remio` or simply `rio`.
> 
> Update: [01c2cbe](https://github.com/bao-project/bao-hypervisor/commit/01c2cbed6e573c28af1b82229e798d1298da16a6)

The namespace change should be reflected in the function, variable names, enums, etc

DavidMCerdeira commented on src/core/remio.c:
```diff
@@ -0,0 +1,668 @@
+ * @param id Remote I/O device ID
+ * @return Returns the Remote I/O device or NULL if the device was not found
+ */
+static struct remio_device* remio_find_vm_io_dev_by_id(struct vm* vm, unsigned long id)
```
I suggest this name, but now realize it should be `remio_find_vm_io_dev_by_id`
other "_io_dev" instances should be changed similarly

joaopeixoto13 replied:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/29c3e19557de3c3aaa1cac56e49b8b535d53fe0c

DavidMCerdeira commented on src/core/remio.c:
```diff
@@ -0,0 +1,668 @@
+                list_push(&remio_device_list, (node_t*)device);
+                backend_cnt++;
+                devices[dev->id][REMIO_DEV_BACKEND] = (int)vm_id;
+            } else {
```
Test for `REMIO_DEV_FRONTEND` explicitly
There are other instances of this, please update them as well

joaopeixoto13 replied:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/b0212dc3c64a3f834cf0c44a92d75b6f0ae354a4

DavidMCerdeira commented on src/core/remio.c:
```diff
@@ -0,0 +1,668 @@
+long int remio_hypercall(unsigned long arg0, unsigned long arg1, unsigned long arg2)
+{
+    long int ret = -HC_E_SUCCESS;
+    unsigned long virt_remio_dev_id = arg0;
```
why "virt"
should be remio_dev_id, or just dev_id

joaopeixoto13 replied:
The `virt` keyword implies that the backend VM is unaware of the absolute/remote I/O ID. For example, if there are two backend VMs, each with two remote I/O devices, the backend will work with devices IDs of 0 and 1. It is the responsibility of the Remote I/O system to calculate the correct absolute ID.

That said, I renamed `abs_remio_dev_id` to `remio_dev_id` for better clarity and alignment with its function.

Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/b79edf26d690526668c2a363f5f979e80fe1ee1e

DavidMCerdeira commented on src/core/vm.c:
```diff
@@ -225,6 +228,55 @@ static void vm_init_dev(struct vm* vm, const struct vm_config* vm_config)
+
+    for (size_t i = 0; i < vm_config->platform.remio_dev_num; i++) {
+        struct remio_dev* remio_dev = &vm_config->platform.remio_devs[i];
+        struct shmem* shmem = shmem_get(remio_dev->shmem.shmem_id);
+        if (shmem == NULL) {
+            WARNING("Invalid shmem id in configuration. Ignored.");
+            continue;
+        }
+        size_t shmem_size = remio_dev->shmem.size;
+        if (shmem_size > shmem->size) {
+            shmem_size = shmem->size;
+            WARNING("Trying to map region to smaller shared memory. Truncated");
+        }
+        spin_lock(&shmem->lock);
+        shmem->cpu_masters |= (1UL << cpu()->id);
+        spin_unlock(&shmem->lock);
+
+        struct vm_mem_region reg = {
+            .base = remio_dev->shmem.base,
+            .size = shmem_size,
+            .place_phys = true,
+            .phys = shmem->phys,
+            .colors = shmem->colors,
+        };
+
+        vm_map_mem_region(vm, &reg);
+
+        if (remio_dev->type == REMIO_DEV_FRONTEND) {
+            struct emul_mem* emu = objpool_alloc(&emul_cache);
+            if (emu == NULL) {
+                ERROR("Failed allocating emulation memory node");
+            }
+            emu->va_base = remio_dev->va;
+            emu->size = remio_dev->size;
+            emu->handler = remio_mmio_emul_handler;
+            vm_emul_add_mem(vm, emu);
+        }
```
move to a new function `vm_init_remio_dev`

joaopeixoto13 replied:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/7e8cd56617b25b56b88306f43ab9c491286c3764

DavidMCerdeira commented on src/core/remio.c:
```diff
@@ -0,0 +1,668 @@
+    }
+
+    /** Find the Remote I/O device based on the Remote I/O device ID */
+    list_foreach (remio_device_list, struct remio_device, io_device) {
+        if (id == io_device->id) {
```
create function to find by id `remio_find_dev_id`
you can use it here and in line 345 if I'm not mistaken

joaopeixoto13 replied:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/a7f81943129b9011cf5eafd47597d7e98307af85

joaopeixoto13 commented on src/core/vm.c:
```diff
@@ -225,6 +228,60 @@ static void vm_init_dev(struct vm* vm, const struct vm_config* vm_config)
     }
 }
 
+static void vm_init_remio_dev(struct remio_dev* remio_dev, struct vm* vm)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/704381b1d4eb6f8b8ea5ee3fd4fc895568f8db78

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,685 @@
+        struct vm_config* vm_config = &config.vmlist[vm_id];
+        for (size_t i = 0; i < vm_config->platform.remio_dev_num; i++) {
+            struct remio_dev* dev = &vm_config->platform.remio_devs[i];
+            struct remio_device* device = remio_find_dev_by_id(dev->id);
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/dec517eee771d789a3e5e8b678e13c012815d537

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,685 @@
+                device->config.backend.vm_id = vm_id;
+                device->config.backend.interrupt = dev->interrupt;
+                device->config.backend.cpu_id = (cpuid_t)-1;
+            } else {
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/f6b868da35fd569beb0f79d50934d6736649c4a8

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,685 @@
+
+void remio_assign_vm_cpus(struct vm* vm)
+{
+    list_foreach (remio_device_list, struct remio_device, io_device) {
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/0c544f9f159dd37e609e0b8516eb508109f5ddaf

DavidMCerdeira reviewed (commented):
I'd still like to test this before approving, but for now it looks good!
:+1:

joaopeixoto13 commented:
> I'd still like to test this before approving, but for now it looks good! 👍 You can squash the commits into a single one now

Thanks, @DavidMCerdeira ! Also, please take a look at this [commit](https://github.com/bao-project/bao-hypervisor/pull/176/commits/bdb870fce0a8c4e0ca82c6c4c9095a142dcdaad7), where I temporarily resolved the issue by replacing the `psci_power_down` call with a `WFI` instruction. However, I believe we need to find a more appropriate long-term solution. I haven't reviewed the current [guest reboot PR](https://github.com/bao-project/bao-hypervisor/pull/157) yet, but it may offer some insight or help in addressing this issue.

DavidMCerdeira commented:
> > I'd still like to test this before approving, but for now it looks good! 👍 You can squash the commits into a single one now
> 
> Thanks, @DavidMCerdeira ! Also, please take a look at this [commit](https://github.com/bao-project/bao-hypervisor/pull/176/commits/bdb870fce0a8c4e0ca82c6c4c9095a142dcdaad7), where I temporarily resolved the issue by replacing the `psci_power_down` call with a `WFI` instruction. However, I believe we need to find a more appropriate long-term solution. I haven't reviewed the current [guest reboot PR](https://github.com/bao-project/bao-hypervisor/pull/157) yet, but it may offer some insight or help in addressing this issue.

I think for now it is fine. Actually we are aware of at least two platforms that require this behavior to work with bao currently. We should eventually allow platforms to have specific code to handle this sort of platform specific quirks

josecm commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+struct list remio_device_list;
+
+/** Array of Remote I/O requests */
+struct remio_request remio_requests[REMIO_CPU_NUM][REMIO_VCPU_NUM];
```
More, I don't really like the way were are allocating requests. I would suggest changing this structure. Since each request is associated with an event, we could merge the two concepts and only free the object from the pool once the request was completed. 

Then instead of indexing requests with the cpu_id/vcpu_id combination we would have a dynamically allocated id for events/requests that would be passed to the backend for later identification on hypercalls. Also associated with it, we would have a "backend owner VM id" that the hypercall handler would use to check that that event/request is indeed being hanlded by that VM.

joaopeixoto13 replied:
> This seems to be wasting memory since for a given CPU there is at most one vcpu.

Yes, you're right. However, I implemented it this way to allow for potential future enhancements in Bao, such as exploring more advanced configurations. These could include 1-to-many mappings between physical and virtual CPUs, or sharing the same vCPU across multiple VMs. In the former case, the vCPU ID would correspond to the actual vCPU, while in the latter, it would be associated with the VM ID.

joaopeixoto13 replied:
> More, I don't really like the way were are allocating requests. I would suggest changing this structure. Since each request is associated with an event, we could merge the two concepts and only free the object from the pool once the request was completed.

From my tests, I attempted to merge the two concepts, but based on my experience, this introduces higher costs in terms of both code complexity and performance. First, since the request event is associated with a specific remote I/O device, when the backend requests a new event, it doesn’t need to iterate through all vCPUs to retrieve it. Instead, it can simply pop the oldest element from the list.

Additionally, the issue of vCPU prioritization that I mentioned earlier still persists when merging the two concepts. Furthermore, as the list is already protected by a spinlock, there’s no need for additional synchronization primitives, which results in simpler code.

joaopeixoto13 replied:
> Then instead of indexing requests with the cpu_id/vcpu_id combination we would have a dynamically allocated id for events/requests that would be passed to the backend for later identification on hypercalls.

But how do you locate the request? Are you suggesting iterating over all vCPUs associated with a frontend VM? The same issue I mentioned earlier would persist in this approach.

I do agree that we could dynamically create a unique identifier to replace the CPU and vCPU IDs. Perhaps we could apply a basic mask and concatenate the two numbers into a single unsigned long variable, generating a unique identifier that could then be used to index the request.

> Also associated with it, we would have a "backend owner VM id" that the hypercall handler would use to check that that event/request is indeed being hanlded by that VM.

I think I may have misunderstood what you referred to as the "backend owner VM ID." However, the current implementation already verifies that the backend VM performing the hypercall is the owner of the remote I/O backend device. Since the event request list is associated with each remote I/O device, we ensure that only the corresponding backend interacts with the event request.

josecm replied:
The logic of handling the request queue would be exactly the same. The only thing that would fundamentally change in my suggestion would be the way you allocate events/requests and the way the backend identifies the specific event/request on an hypercall.

joaopeixoto13 replied:
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/3e9940ee1b0566c3ad0b9049454f3e61f7adbc1b

josecm replied:
I still think we dont need the event array in the device. We can directly access it via the objpool. We just need some check to make sure that event belongs to the VM calling the ask hypercall and that that VM is the backend for the device associated with that event.

joaopeixoto13 replied:
Solved [here](https://github.com/bao-project/bao-hypervisor/pull/176/commits/77ff8e893f35c268097812e29bcc0463c8796d43)

josecm reviewed (changes requested):
Despite the very (!) good work, I think we still need a bit of work on this PR. This is my first pass on it. I think we will need one or two more.

joaopeixoto13 commented on src/arch/armv8/armv8-a/cpu.c:
```diff
@@ -29,7 +29,7 @@ void cpu_arch_profile_init(cpuid_t cpuid, paddr_t load_addr)
 void cpu_arch_profile_idle()
 {
-    int64_t err = psci_power_down(PSCI_WAKEUP_IDLE);
+    int64_t err = PSCI_E_NOT_SUPPORTED;
```
I may have misunderstood your message. Has the issue already been resolved, or do you suggest separating this [commit](https://github.com/bao-project/bao-hypervisor/pull/176/commits/bdb870fce0a8c4e0ca82c6c4c9095a142dcdaad7) into a new PR and making this current PR dependent on the former one?

josecm replied:
This issue has been solved by @DavidMCerdeira in PR #176, I believe.

joaopeixoto13 replied:
I found the misunderstanding. The PR hyperlink is incorrect. The fix is in this [PR](https://github.com/bao-project/bao-hypervisor/pull/179), not in [this one](https://github.com/bao-project/bao-hypervisor/pull/176), which is the current PR.

josecm replied:
Yes, sorry for the confusion!

joaopeixoto13 commented on src/arch/riscv/sync_exceptions.c:
```diff
@@ -152,5 +152,8 @@ void sync_exception_handler(void)
     }
 
-    cpu()->vcpu->regs.sepc += pc_step;
+    vcpu_writepc(cpu()->vcpu, vcpu_readpc(cpu()->vcpu) + pc_step);
```
Yes, you're correct. However, I decided to use the generic function to maintain consistency with the approach we have in the `src/arch/armv8/aborts.c` file. Although it's an architecture-specific file, it still utilizes this generic function.

josecm replied:
OK. Not really necessary, but we can maintain this.

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+    unsigned long virt_remio_dev_id = arg0;
+    unsigned long addr = arg1;
+    unsigned long op = arg2;
+    unsigned long value = vcpu_readreg(cpu()->vcpu, HYPCALL_IN_ARG_REG(3));
+    unsigned long cpu_id = vcpu_readreg(cpu()->vcpu, HYPCALL_IN_ARG_REG(4));
+    unsigned long vcpu_id = vcpu_readreg(cpu()->vcpu, HYPCALL_IN_ARG_REG(5));
```
If you agree, I can add a commit to modify the `hypercall` function to retrieve additional arguments (e.g., three in this case).

josecm replied:
Ideally we would come up with a way of registering hypercall handlers with different numbers of arguments...

DavidMCerdeira replied:
I think adding an API for reading arguments is a good idea
This way each handler parses the arguments as it sees fit

josecm replied:
@joaopeixoto13 did you address this?

joaopeixoto13 replied:
Now in this [commit](https://github.com/bao-project/bao-hypervisor/pull/176/commits/0eba03f7ad9932fb38d6db1c0611c0eece14ff5c)

josecm commented on src/arch/riscv/sbi.c:
```diff
@@ -405,6 +405,7 @@ static struct sbiret sbi_bao_handler(unsigned long fid)
     struct sbiret ret;
 
     ret.error = hypercall(fid);
+    ret.value = (long int)vcpu_readreg(cpu()->vcpu, REG_A1);
```
I don't get your point. I don't see any other ecall handler function that does that.

joaopeixoto13 replied:
I was referring to the `sbi_vs_handler` function, which already uses `ret.value` to write to the `REG_A1` register in the final line before the return statement. Instead of removing this line, I read the value into the `sbiret` ret struct first to prevent it from being overwritten. This is necessary because writing to the register in the `remio_handle_ask` function would overwrite the same value if `vcpu_writereg(cpu()->vcpu, REG_A1, (unsigned long)ret.value)` remains.

joaopeixoto13 replied:
> I see your point. It seems we are actually breaking the [SBI spec](https://github.com/riscv-non-isa/riscv-sbi-doc/blob/master/src/binary-encoding.adoc), as it states "All registers except a0 & a1 must be preserved across an SBI call by the callee.". There is no way around this as the alternative would be using a pointer to some shared memory area which is something we really don't want. We should make sure a1 contains a value part of Bao's hypercall error codes (return status as you mention). Then the remaining return values are returned following the logic of the normal RISC-V ABI (a2, a3, etc...). So maybe we should change this macro to `#define HYPCALL_OUT_ARG_REG(ARG) (HYPCALL_IN_ARG_REG((ARG) + 2))`. What do you think?

Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/47f09052fc338b4e458bb60f3f66bf7c20ad5b25

josecm commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+    switch (op) {
+        case REMIO_HYP_WRITE:
+        case REMIO_HYP_READ:
+            if (!remio_handle_rw(value, cpu_id, vcpu_id)) {
```
But this is not checked on the `remio_handle_rw` call. The `cpu_id`/`vcpu_id` are used to access the request without any check.

joaopeixoto13 replied:
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/3e9940ee1b0566c3ad0b9049454f3e61f7adbc1b
After validating and extracting the associated Remote I/O device from the backend VM, we now pass the device pointer to the `remio_handle_rw`. Since the requests are tied to the specific Remote I/O device, this ensures that only the correct device on the correct backend VM can modify its state.

josecm replied:
There is still an issue here. Calling `remio_handle_rw` will end up in calling `remio_get_request` that performs the access to the request array without checking if the id is valid!

joaopeixoto13 replied:
Done [here](https://github.com/bao-project/bao-hypervisor/pull/176/commits/77ff8e893f35c268097812e29bcc0463c8796d43)

joaopeixoto13 commented on src/core/inc/vm.h:
```diff
@@ -84,6 +88,9 @@ struct vm {
     size_t ipc_num;
     struct ipc* ipcs;
+
+    size_t remio_dev_num;
+    struct remio_dev* remio_devs;
 };
 
```
This is not necessarily mandatory. However, it simplifies the process of extracting Remote I/O devices in both the `remio_find_vm_dev_by_id` and `remio_find_vm_dev_by_addr` methods.

However, I'm curious why you applied the same approach for the IPCs, as all the necessary information can also be obtained from the config structure. 

If you're okay with it, I can remove the Remote I/O devices from the VM, as well as the IPCs, to maintain consistency.

joaopeixoto13 replied:
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/4e6a971d7dc31dc9dd725c59ce781e83575e7ef5

josecm replied:
You are spot on in your comparison! But you didn't change the IPC stuff. Maybe leave it as you had before. (Sorry)

joaopeixoto13 commented on src/core/inc/remio.h:
```diff
@@ -0,0 +1,87 @@
+#ifndef __REMIO_H__
+#define __REMIO_H__
+
+#include "types.h"
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/2bc6937f1c0ff646269b97e642dac899125c879f

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ */
+
+#include "inc/bao.h"
+#include "inc/types.h"
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/2bc6937f1c0ff646269b97e642dac899125c879f

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+#include <objpool.h>
+#include <config.h>
+#include <spinlock.h>
+#include <util.h>
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/2bc6937f1c0ff646269b97e642dac899125c879f

joaopeixoto13 commented on src/core/vm.c:
```diff
@@ -9,6 +9,9 @@
 #include <shmem.h>
+#include <objpool.h>
+
+OBJPOOL_ALLOC(emul_cache, struct emul_mem, sizeof(struct emul_mem));
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/01758417bf19ede7cce6d33e62fda929400133a9

josecm replied:
I don't think you got my point. I am saying that there is no need for the pool itself. You can allocate this in the config stucture.

josecm replied:
Or if you want to keep it in the pool it is also ok. Your call.

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+OBJPOOL_ALLOC(remio_device_pool, struct remio_device, sizeof(struct remio_device));
+
+/** Object pool to allocate pending Remote I/O requests events */
+OBJPOOL_ALLOC(remio_request_event_pool, struct remio_request_event,
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/01758417bf19ede7cce6d33e62fda929400133a9

joaopeixoto13 commented on src/core/vm.c:
```diff
@@ -9,6 +9,9 @@
 #include <shmem.h>
+#include <objpool.h>
+
+OBJPOOL_ALLOC(emul_cache, struct emul_mem, sizeof(struct emul_mem));
```
Done https://github.com/bao-project/bao-hypervisor/commit/01758417bf19ede7cce6d33e62fda929400133a9

joaopeixoto13 commented on src/core/inc/remio.h:
```diff
@@ -0,0 +1,87 @@
+#ifndef __REMIO_H__
+#define __REMIO_H__
+
+#include "types.h"
```
Done https://github.com/bao-project/bao-hypervisor/commit/2bc6937f1c0ff646269b97e642dac899125c879f

joaopeixoto13 commented on src/core/inc/remio.h:
```diff
@@ -0,0 +1,87 @@
+ */
+
+#ifndef __REMIO_H__
+#define __REMIO_H__
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/0af3be6640a68a068bff2fde4f175dda1c1b9519

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns the Remote I/O request
+ */
+static struct remio_request* remio_get_request(unsigned long cpu_id, unsigned long vcpu_id)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @param request Pointer to the Remote I/O request
+ */
+static void remio_insert_request(unsigned long cpu_id, unsigned long vcpu_id,
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @param state New state of the I/O request
+ */
+static void remio_set_request_state(unsigned long cpu_id, unsigned long vcpu_id,
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @return Returns the state of the I/O request
+ */
+static enum REMIO_STATE remio_get_request_state(unsigned long cpu_id, unsigned long vcpu_id)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param vcpu_id vCPU ID of the frontend VM that issued the I/O request
+ * @param value New value of the I/O request
+ */
+static void remio_set_request_value(unsigned long cpu_id, unsigned long vcpu_id, unsigned long value)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param device Pointer to the Remote I/O device
+ * @param event Pointer to the Remote I/O request event
+ */
+static void remio_push_request_event(struct remio_device* device, struct remio_request_event* event)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param device Pointer to the Remote I/O device
+ * @return Returns the next pending I/O request event or NULL if there is no pending I/O request
+ */
+static struct remio_request_event* remio_pop_request_event(struct remio_device* device)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param device Pointer to the Remote I/O device
+ * @return Returns the number of pending I/O requests
+ */
+static size_t remio_get_request_event_count(struct remio_device* device)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/1bf100a5ad62550862f5ea4bf48c35409da3bdcd

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+
+/** Object pool to allocate pending Remote I/O requests events */
+OBJPOOL_ALLOC(remio_request_event_pool, struct remio_request_event,
+    sizeof(struct remio_request_event));
```
Done https://github.com/bao-project/bao-hypervisor/commit/01758417bf19ede7cce6d33e62fda929400133a9

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ * @param acc Pointer to the emul_access structure containing the MMIO access information
+ * @return Returns the Remote I/O request
+ */
+static struct remio_request remio_create_request(struct emul_access* acc)
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/9d34ff2ad5ca031a25388ca26deaf9a98f459637

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+ */
+static struct remio_request_event* remio_create_event(void)
+{
+    struct remio_request_event* event = objpool_alloc(&remio_request_event_pool);
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/3e9940ee1b0566c3ad0b9049454f3e61f7adbc1b

joaopeixoto13 commented on src/core/remio.c:
```diff
@@ -0,0 +1,709 @@
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(4), event->cpu_id);
+    vcpu_writereg(cpu()->vcpu, HYPCALL_OUT_ARG_REG(5), event->vcpu_id);
+
+    objpool_free(&remio_request_event_pool, event);
```
Done https://github.com/bao-project/bao-hypervisor/pull/176/commits/3e9940ee1b0566c3ad0b9049454f3e61f7adbc1b

josecm commented:
@joaopeixoto13, to keep a clean history, try as much as possible remove the "fix" commits and introduce the change correctly in the first commit that code is introduced.

joaopeixoto13 commented:
> @joaopeixoto13, to keep a clean history, try as much as possible remove the "fix" commits and introduce the change correctly in the first commit that code is introduced.

Yes, I understand. I split these smaller fixes to make it easier for you to review the changes. At the end of the PR, I’ll rebase all the fix commits into the main "parent" commit that introduced the feature.

DavidMCerdeira commented on src/core/remio.c:
```diff
@@ -0,0 +1,713 @@
+    return true;
+}
+
+static void remio_cpu_handler(uint32_t event, uint64_t data)
```
```suggestion
static void remio_cpu_msg_handler(uint32_t event, uint64_t data)
```

joaopeixoto13 replied:
Renamed [here](https://github.com/bao-project/bao-hypervisor/pull/176/commits/6a55d0e46b161a6104d372bf4739a2aeeaf1085b).

DavidMCerdeira commented on src/core/inc/config_defs.h:
```diff
@@ -11,6 +11,7 @@
 #define CONFIG_VCPU_NUM      1
 #define CONFIG_VM_NUM        1
 #define CONFIG_HYP_BASE_ADDR 0
+#define CONFIG_REMIO_DEV_NUM 0
```
If this value is set 1, there's no need for `CONFIG_REMIO_DEV_NUM ? CONFIG_REMIO_DEV_NUM : 1` in remio.c and vm.c

joaopeixoto13 replied:
> If this value is set 1, there's no need for `CONFIG_REMIO_DEV_NUM ? CONFIG_REMIO_DEV_NUM : 1` in remio.c and vm.c

The issue here arises when no remote I/O device is selected by the user, causing the OBJPOOL_ALLOC to be sized at 0. This leads to an error due to ISO C's prohibition on zero-size arrays.

DavidMCerdeira replied:
I was actually suggesting setting 1 in the script, because it bothers this logic is performed in two places.
But it's not too bad, and the REMIO_DEV_NUM would lose some of its meaning
It's okay as is

DavidMCerdeira reviewed (approved):
Only nitpicks
Otherwise looks good to me!

josecm commented:
@joaopeixoto13 rebase on main and resolve conflicts please

joaopeixoto13 commented:
> @joaopeixoto13 rebase on main and resolve conflicts please

Done

josecm commented:
> Done. If everything looks good, I can proceed to merge all the fix commits into the main parent commit that introduced the feature.

Please go ahead with that.

joaopeixoto13 commented:
> > Done. If everything looks good, I can proceed to merge all the fix commits into the main parent commit that introduced the feature.
> 
> Please go ahead with that.

Done.

josecm commented on src/core/inc/hypercall.h:
```diff
@@ -8,14 +8,24 @@
 
 long int hypercall(unsigned long id);
 
+inline unsigned long hypercall_get_arg(struct vcpu* vcpu, size_t arg_index)
```
```suggestion
static inline unsigned long hypercall_get_arg(struct vcpu* vcpu, size_t arg_index)
```

josecm commented on src/core/inc/hypercall.h:
```diff
@@ -8,14 +8,24 @@
+    return vcpu_readreg(vcpu, HYPCALL_IN_ARG_REG(arg_index));
+}
+
+inline void hypercall_set_ret(struct vcpu* vcpu, size_t arg_index, unsigned long arg_val)
```
```suggestion
static inline void hypercall_set_ret(struct vcpu* vcpu, size_t arg_index, unsigned long arg_val)
```

josecm commented on src/core/remio.c:
```diff
@@ -0,0 +1,641 @@
+    }
+
+    /** Calculate the remaining number of pending I/O requests */
+    pending_requests = (long unsigned int)remio_num_pending_request(device);
```
Is this really necessary? Can't we just return some flag that says that there are more requests waiting? If so we don't need to go through the list to count its size in list_size. Just use list_emtpy...

josecm replied:
I think it was the right way to go to remove it.
